mapswipe_workers/mapswipe_workers/firebase_to_postgres/transfer_results.py

```diff
@@ -55,24 +42,92 @@ def transfer(current_results):
             continue
+        else:
+            transfer_results_for_project(project_id)
+
+    return project_id_list
+
+
+def transfer_results_for_project(project_id):
+    """Transfer the results for a specific project.
+
+    Download all results for a project from firebase.
+    Save results into an in-memory file.
+    Copy the results to postgres.
+    Delete results in firebase.
+
+    We are NOT using a Firebase transaction functions here anymore.
+    This has caused problems, in situations where a lot of mappers are
+    uploading results to Firebase at the same time. Basically, this is
+    due to the behaviour of Firebase Transaction function:
+
+        "If another client writes to this location
+        before the new value is successfully saved,
+        the update function is called again with the new current value,
+        and the write will be retried."
 
-        logger.info(f"{project_id}: Start transfering results")
+    Using Firebase transaction on the group level
+    has turned out to be too slow when using "normal" queries,
+    e.g. without using threading. Threading should be avoided here
+    as well to not run into unforeseen errors.
 
-        results_ref = fb_db.reference(f"v2/results/{project_id}")
+    For more details see issue #478.
+    """
+    # TODO: which exceptions can happen here?
+
+    logger.info(f"{project_id}: Start transfer results")
+    fb_db = auth.firebaseDB()
+    results_ref = fb_db.reference(f"v2/results/{project_id}")
+    results = results_ref.get()
+
+    if results is None:
+        logger.info(f"{project_id}: No results in Firebase")
+    else:
+        # First we check for new users in Firebase.
+        # The user_id is used as a key in the postgres database for the results
+        # and thus users need to be inserted before results get inserted.
+        results_user_id_list = get_user_ids_from_results(results)
+        update_data.update_user_data(results_user_id_list)
+
+        # Results are dumped into an in-memory file.
+        # This allows us to use the COPY statement to insert many
+        # results at relatively high speed.
+        results_file = results_to_file(results, project_id)
         truncate_temp_results()
+        save_results_to_postgres(results_file)
 
-        try:
-            results_ref.transaction(transfer)
-            logger.info(f"{project_id}: Transfered results to postgres")
-        except fb_db.TransactionAbortedError:
-            logger.exception(
-                f"{project_id}: Firebase transaction for "
-                f"transfering results failed to commit"
-            )
-            sentry.capture_exception()
+        # It is important here that we first insert results into postgres
+        # and then delete these results from Firebase.
+        # In case something goes wrong during the insert, results in Firebase
+        # will not get deleted.
+        delete_results_from_firebase(project_id, results)
+        del fb_db
```

Comment:
Any special reason you delete `fb_db` here? it will get deleted automatically next line, when it becomes out of scope.

Reply:
no, removed